[14.0] Add shopfloor_package_measurement

TDu opened the pull request:
Depends on:
- [x] https://github.com/OCA/delivery-carrier/pull/522

ref: rau-43

simahawk commented on shopfloor_checkout_package_measurement/actions/schema.py:
```diff
@@ -0,0 +1,18 @@
+    _inherit = "shopfloor.schema.action"
+
+    def package(self, with_packaging=False):
+        res = super().package(with_packaging)
```
```suggestion
        res = super().package(with_packaging=with_packaging)
```

TDu replied:
Changed

simahawk commented on shopfloor_checkout_package_measurement/services/checkout.py:
```diff
@@ -0,0 +1,116 @@
+            return self._response_for_summary(
+                picking, message=self.msg_store.record_not_found()
+            )
+        if height:
```
cant we get a single dict called `measures` instead of single values? Not important tho

simahawk commented on shopfloor_checkout_package_measurement/services/checkout.py:
```diff
@@ -0,0 +1,116 @@
+            picking,
+            message={
+                "message_type": "success",
+                "body": _("Package measurement changed."),
```
to be moved to action messages

TDu replied:
done

simahawk commented on shopfloor_checkout_package_measurement/tests/test_checkout_package_measurement.py:
```diff
@@ -0,0 +1,126 @@
+        cls.picking.action_assign()
+
+    def test_requested_measurement_for_new_package(self):
+        """Check package measurement are requested fro new package."""
```
```suggestion
        """Check package measurement are requested for new package."""
```

TDu replied:
fixed

simahawk commented on shopfloor_checkout_package_measurement_mobile/static/wms/src/scenario/checkout_package_measurement.js:
```diff
@@ -0,0 +1,100 @@
+
+const checkout_scenario = process_registry.get("checkout");
+const template = checkout_scenario.component.template;
+const pos = template.indexOf("</Screen>");
```
A comment explaining what you do and why would be nice.

TDu replied:
Added

simahawk commented:
@TDu remember to add the dependency in the description please.

simahawk commented on shopfloor_checkout_package_measurement_mobile/static/wms/src/scenario/checkout_package_measurement.js:
```diff
@@ -0,0 +1,100 @@
+            },
+            events: {
+                go_back: "on_back",
+                done: "on_done",
```
This is useless if you don't trigger the event `done`.
You are using `@click` above -> no need to bind again here AFAIS.

TDu replied:
Removed

simahawk reviewed (changes requested):
Also, please, put core changes in their own commits.

TDu committed:
shopfloor: add checkout scenario uml graph

TDu committed:
Add shopfloor_checkout_package_measurement

TDu committed:
shopfloor_mobile_base: fix registry replace

TDu committed:
shopfloor_mobile: allow extenstion of states on checkout

This change allows for other modules to extend the state of the checkout
scenario.
See the `shopfloor_checkout_package_measurement_mobile` for a use of
this change.

TDu committed:
Add shopfloor_checkout_package_measurement_mobile

TDu commented:
I have split the commit for the change in the base addons.
Also the fixups have the change requested as well as some new improvements (only display the required measurement, add the uom and use the shipping_weight field.

simahawk commented on shopfloor_checkout_package_measurement/actions/message.py:
```diff
@@ -0,0 +1,15 @@
+    def package_measurement_changed(self):
+        return {
+            "message_type": "success",
+            "body": _("Package measurement changed."),
```
```suggestion
            "body": _("Package measure(s) changed."),
```

TDu replied:
done

simahawk replied:
you didn't update tests ;)

simahawk commented on shopfloor_checkout_package_measurement_mobile/static/wms/src/scenario/checkout_package_measurement.js:
```diff
@@ -0,0 +1,114 @@
+/**
```
can you remove `wms` from the path pls?

TDu replied:
done

simahawk commented on shopfloor_mobile/static/wms/src/scenario/checkout_states.js:
```diff
@@ -0,0 +1,366 @@
+/**
+ * Copyright 2022 Camptocamp SA (http://www.camptocamp.com)
```
```suggestion
 * Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
```

TDu replied:
done

simahawk commented on shopfloor_mobile/static/wms/src/scenario/checkout_states.js:
```diff
@@ -0,0 +1,366 @@
+ * License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+ */
+
+export const checkout_states = function ($instance) {
```
```suggestion

/*
    Define states for checkout scenario.
    @param $instance VueJS component instance  
*/
export const checkout_states = function ($instance) {
```

TDu replied:
done

simahawk reviewed (commented):
LG, keeping blocked till dep is merged

jbaudoux commented on shopfloor_checkout_package_measurement_mobile/static/src/scenario/checkout_package_measurement.js:
```diff
@@ -0,0 +1,114 @@
+    <v-form>
+        <v-container>
+            <v-row>
+                <v-text-field
```
If I remember properly, v-text-field does not manage properly the display of the virtual keyboard on a mobile device. This should likely be replaced by the searchbar or something alike and you should set the expected type, in this case number, to show the right keyboard

TDu replied:
The searchbar uses a v-text-field as well. Oh no wait, are you refering to this PR https://github.com/OCA/wms/pull/350/files I see it is not yet merged in 14.0 and the `v-text-field` has been replaced by an `input`element ?

TDu replied:
The problem is by switching to the `input` element we will loose all the nice thing of Material Design and vuetify and the rendering of the field will not  be nice at all.
Do we know why both are not compatible ?

jbaudoux replied:
@TDu As we have to type something in the field, we don't have to hide the virtual keyboard. So fine to use v-text-field.
Still need to put type number to display right keyboard and prevent to enter chars

TDu replied:
Done

jbaudoux reviewed (changes requested):
Mobile device Virtual keyboard (Gboard) display issue. Can you test from an android device?

jbaudoux commented:
Yes indeed. Should be merged by the way

jbaudoux commented:
Add some space on the screen between the fields:
![image](https://user-images.githubusercontent.com/7802725/194321107-cf59b2b2-3fd7-4e86-a558-3f8c95a8d364.png)

TDu commented:
> Add some space on the screen between the fields: ![image](https://user-images.githubusercontent.com/7802725/194321107-cf59b2b2-3fd7-4e86-a558-3f8c95a8d364.png)

Done

jbaudoux commented:
@simahawk Can you update your review?

simahawk commented on shopfloor_checkout_package_measurement/services/checkout.py:
```diff
@@ -0,0 +1,125 @@
+
+    def _put_lines_in_allowed_package(self, picking, selected_lines, package):
+        res = super()._put_lines_in_allowed_package(picking, selected_lines, package)
+        if not picking._check_required_package_measurement(package=package):
```
this line has to be adapted to changes in https://github.com/OCA/delivery-carrier/pull/522

TDu replied:
Yes. But actually the latest changes done in the related PR actually break this one in more ways...
Now that we have the required measurements checked by a constrain this code does not work anymore, because we first save the stock.quant.package (the constrain stops us here) and then check if some of the measurements are required,
One way to get around this would be 
```
    @api.constrains("length", "width", "height", "weight")
    def _check_required_dimension(self, ignore_picking_state=False):
        for package in self:
            domain = [('result_package_id', '=', package.id)]
            pickings = self.env['stock.move.line'].search(domain).mapped('picking_id')  
            if pickings.state != "done" and not ignore_picking_state:
                continue
            required_dimension = []
            if package.length_required and not package.pack_length:
                required_dimension.append(_("length"))
...
```
It can be improve with a read_group but still not that nice...

TDu replied:
Here is the PR https://github.com/OCA/delivery-carrier/pull/546 to fix that issue...

TDu commented:
I squashed all the reviewed fixes. And updated the pr to handle the new changes on https://github.com/OCA/delivery-carrier/pull/546

simahawk commented on shopfloor_checkout_package_measurement/services/checkout.py:
```diff
@@ -61,7 +63,11 @@ def set_package_measurement(
         res = super()._put_lines_in_allowed_package(picking, selected_lines, package)
-        if not picking._check_required_package_measurement(package=package):
+        try:
+            package.with_context(
```
See my comment on the other PR.
Also, never override the whole ctx unless you really want to. Use kw arg instead.

TDu replied:
fixed

TDu commented:
rebased

TDu commented:
> LG, keeping blocked till dep is merged

@simahawk dependency has been merged...

simahawk commented on shopfloor_checkout_package_measurement/actions/schema.py:
```diff
@@ -0,0 +1,37 @@
+
+    def package(self, with_packaging=False):
+        res = super().package(with_packaging=with_packaging)
+        res.update(
```
doesn't make sense only if `with_packaging` is True?

TDu replied:
I would say no, because the measurements are for packages and not for packaging.

simahawk reviewed (approved):
Minor question

sebalix reviewed (approved):
LG, fixup to squash and we can merge

simahawk commented:
/ocabot merge patch
